### ISSUE TITLE
Expose decompression_executor in uproot.dask

### ISSUE THREAD

**Issue by NJManganelli** [scikit-hep/uproot5#1079]
In scenarios where multiple dask workers are running on a machine (for example, a local cluster with O(10) workers), only one dask worker will be busy deserializing data from root; multiple threads could be employed in parallel while those other workers are idle/waiting. Especially relevant for dask-localcluster setups

following from this discussion:
https://iris-hep.slack.com/archives/C03UA418VJ8/p1704815356020659

**Comment by nsmith-** [comment]
Why would only one worker be busy? The default TrivialExecutor will run in the current thread and provide work to each dask-worker available.

**Comment by jpivarski** [comment]
I don't entirely understand your comment. This PR propagates the executors into `uproot.dask` so that you can have nested parallelism: Dask parallelizes the partitions (maybe) and `decompression_executor` and `interpretation_executor` do low-level parallelization within that to parallelize decompression and interpretation.

The inner level of parallelism isn't often useful, though the decompression functions and AwkwardForth both release the GIL, so there are some corner cases in which they're useful. For that reason (and for things to be easy to guess/predict), the default executors are TrivialExecutor, which doesn't parallelize.

We're probably in agreement that most of the time, users should parallelize with Dask on multiple processes if a job is big enough to need any parallelization. The one corner-case I can think of in which `decompression_executor` might help is if the file is LZMA-compressed. Up to about 4 threads or so, it scales.

**Comment by nsmith-** [comment]
The issue I'm raising is that, say you ask dask to compute using 4 threads. There's an implicit assumption here that those threads will mostly be saturated with CPU-bound work and your request for a number of threads is a resource usage request. If each of those threads runs a task that involves opening and reading a file with uproot, which in turn provisions an executor with, say, 4 threads, you're bypassing the implicit resource request and have the potential to saturate up to 16 CPU cores, if that many are available. The idea is that we should be assuming that dask is in charge of managing resource usage and each task should be (approximately) single-threaded.
If we need to parallelize decompression across multiple cores (i.e. if we decide the embarassingly-parallel event chunk dimension is not sufficient and we need to parallelize on the basket decompression) then we should expose these decompression workloads to the dask scheduler as separate tasks.

**Comment by jpivarski** [comment]
Yes, nested parallelism is problematic: one should not engage in it without carefully tuning. Are you saying that just providing the knobs is a problem because someone might use them carelessly?

One way that it could be done carefully is to have Dask strictly in charge of multiprocess parallelism and have the executors strictly in charge of multithreaded, in-process parallelism. The executors are only good for the latter, and we generally use Dask for multiprocess only, since Dask has Pythonic overhead that locks the GIL. The Dask workers would have to be launched with `-nthreads 1`.

**Comment by nsmith-** [comment]
Yes I'm concerned the knobs will be turned when they shouldn't be. But maybe it is sufficient to put a disclaimer somewhere in the documentation.

The latter suggestion would mean we don't get to use all the CPU cores available for downstream dask-awkward computations, which don't have any in-built parallelization.

**Comment by jpivarski** [comment]
I've put it on my list[^1]: #1165.

[^1]: I plan to go on a documentation hunting-party soon, aided by ChatGPT. I haven't found time to add documentation on some basic things in years, so I'll fill out the majority of it with AI-generated text and then test all examples and edit mercilessly. The final state won't be a hallucination.
